### ISSUE TITLE
Learner properties of "regr.xgboost" not updated

### ISSUE THREAD

**Issue by Ishant1589** [mlr-org/mlr#978]
Hi,

I am trying to use MLR package for xgboost regression and my data has **missing data points**.

The learner i have used is "regr.xgboost"

Xgboost has the capability to handle missing values but train function of MLR gives me an error that data has missing values and learner does not support that. I am using mlr package instead of xgboost directly because i need partial dependence plots which is not available in xgboost package.

On going deep into the function i discovered that properties associated with learner are only mentioned as - **numeric, factors and weights**.

i edited the function using trace command and added "missings" property and now the function is working fine.

Let me know if MLR package is built in a way to not allow missing data, also if there is any alternative way of running xgboost on missing data using MLR package.

**Comment by berndbischl** [comment]
thx for the report.

such things can happen and we therefore value these reports. i will check what you said and try to update on GH now.


> Let me know if MLR package is built in a way to not allow missing data, also if there is any alternative way of running xgboost on missing data using MLR package.

mlr definitely supports learners with missing data! that is what the property "missing" of a learner is good for. you can even search for such learners, either in the tutorial appendix (learner table) or even directly in R with listLearners and requiring the property missing.

mlr's other current capability to handle missing data is imputing them to allow non-NA-supporting ML learners.

**Comment by berndbischl** [comment]
so if my training dataframe contains NAs I still have to tell xgboost how to detect them? by setting its missing argument to NA? like xgboost(, missing = NA)?

**Comment by berndbischl** [comment]
PR is here #979

**Comment by vinaywv** [comment]
@berndbischl,
usually we provide missing values when creating the DMatrix as below:
dtrain = xgb.DMatrix(data, label=label, missing = -999)

**Comment by berndbischl** [comment]
fixed

**Comment by fabulthuis** [comment]
Hi, 
I see that this issue is closed, but I am trying to use xgboost for regression and clasisfication with NA data points. I have no NaN in my data. 

with learner "regr.xgboost" and "classif.xgboost" I get the error + traceback: 

 Error in xgb.DMatrix(data, label = label, missing = missing) : 
  There are NAN in the matrix, however, you did not set missing=NAN 

 xgb.DMatrix(data, label = label, missing = missing) 
9 xgb.get.DMatrix(data, label, missing) 
8 xgboost::xgboost(data = data, label = target, objective = obj, 
    ...) at RLearner_classif_xgboost.R#44
7 trainLearner.classif.xgboost(.learner = structure(list(id = "classif.xgboost", 
    type = "classif", package = "xgboost", properties = c("twoclass", 
    "multiclass", "numerics", "factors", "prob", "weights", "missings", 
    "featimp"), par.set = structure(list(pars = structure(list( ... at trainLearner.R#23
6 (function (.learner, .task, .subset, .weights = NULL, ...) 
{
    UseMethod("trainLearner")
})(.learner = structure(list(id = "classif.xgboost", type = "classif",  ... 
5 do.call(trainLearner, pars) 
4 fun2(do.call(trainLearner, pars)) 
3 fun1(learner.model <- fun2(do.call(trainLearner, pars))) 
2 system.time(fun1(learner.model <- fun2(do.call(trainLearner, 
    pars))), gcFirst = FALSE) at train.R#31
1 train(lrn_xgboost, task) 

Am I doing something wrong here or is the solution to handle missings not yet implemented.

**Comment by Ishant1589** [comment]
Hi @berndbischl ,

We were facing similar issues with the null data through MLR package. xgboost can only handle NULL data if you explicitly provide the missing property in function 'xgb.DMatrix' by default it is NA and you need to change it to NaN for handling null data.
As we couldn't find any method to do this directly we updated the local copy of the package using trace function. Please find the code below:
```
library (xgboost)
trace("xgb.DMatrix", edit = T)

```
![image](https://cloud.githubusercontent.com/assets/16971666/18306874/7736461c-750b-11e6-98e7-ee5ea866fbd2.png)

Update the default property of missing:
![image](https://cloud.githubusercontent.com/assets/16971666/18306904/a46fde72-750b-11e6-9e71-28a75583bcbc.png)

Save the code using save button and you your code should work for this instance.

**Comment by berndbischl** [comment]
@Ishant1589 @fabulthuis 

Hi guys. They seem to have changed this back forth in xgboost and I have a hard time understanding the xgboost docs here.

Can you PLEASE in cases like these post reproducible examples? We are unit testing xboost in travis, very certainly also with data without NAs or NaNs and that works. So now I am not sure how to help you best.

**Comment by berndbischl** [comment]
and please also open new issues, or reopen this, otherwise this can get lost.

but please really post full, reproducible examples

**Comment by fabulthuis** [comment]
@berndbischl 
@Ishant1589 Can you reopen this issue? 

Here is a reproducible example: 
- create data with missings
- create classification task
- create learner "classif.xgboost"
- train -> error

library(data.table)
library(mlr)
data <- data.table(v1 = rnorm(10, 0, 1),
                   V2 = rnorm(10, 1, 1),
                   V3 = rnorm(10, 4, 2),
                   V4 = rnorm(10, 20, 5),
                   V5 = rnorm(10, 0, 3),
                   v6 = rbinom(10, 1, 0.5))

set(data, 3L, 1L, value=NA)
set(data, 9L, 4L, value=NA)

data = data.frame(data)
data$v6 <- factor(data$v6)

task = makeClassifTask(id = "test", data = data, target = "v6") 

learner = makeLearner("classif.xgboost",                     
                      predict.type = "prob", 
                      par.vals = list(booster = 'gbtree',         
                                      eta = 0.33,
                                      nrounds = 50, 
                                      silent = 0, 
                                      gamma = 4, 
                                      max_depth = 6, 
                                      min_child_weight = 3,
                                      subsample = 0.84,
                                      colsample_bytree = 0.6,
                                      num_parallel_tree = 2,
                                      lambda = 0.7,
                                      lambda_bias = 0.74, 
                                      alpha = 0, 
                                      objective = "binary:logistic",
                                      eval_metric = 'auc',  
                                      base_score = 0.5,
                                      missing = 0,            # also tried NA and NaN
                                      nthread = 1,
                                      feval = NULL, 
                                      verbose = 1, 
                                      print.every.n = 1,
                                      early.stop.round = 16,
                                      maximize = FALSE), 
                      fix.factors.prediction = TRUE)

model = train(learner, task)

The result is the following error:
Error in checkLearnerBeforeTrain(task, learner, weights) : 
  Task 'test' has missing values in 'v1, V4', but learner 'classif.xgboost' does not support that!
